napi_js_props.md

```diff
@@ -0,0 +1,437 @@
+# N-API - JavaScript Data Types and Values 
+
+N-API exposes a set of APIs to get and set properties on JavaScript 
+values. Some of these types are documented under 
+[Section 7](https://tc39.github.io/ecma262/#sec-operations-on-objects) of the 
+[ECMAScript Language Specification](https://tc39.github.io/ecma262/).
+
+Properties in JavaScript are represented as a tuple of a key and a value. 
+Fundamentally, all property keys in N-API can be represented in one of the
+following forms:
+- Named: a simple utf8 encoded string
+- Integer-Indexed: an index value represented by `uint32_t`
+- JavaScript value: these are represented in N-API by `napi_value`. This can
+be a `napi_value` representing a String, Number or Symbol.
+
+N-API values are represented by the type `napi_value`.
+Any N-API call that requires a JavaScript value takes in a `napi_value`.
+However, it's the caller's responsibility to make sure that the 
+`napi_value` in question is of the JavaScript type expected by the API.
+
+The APIs documented in this section provide a simple interface to allow you to 
+get and set properties on arbitrary JavaScript objects represented by 
+`napi_value`.
+
+For instance, consider the following JavaScript code snippet:
+```js
+var obj = {};
+obj.myProp = 123;
+```
+You can do the equivalent using N-API values with the following snippet:
+```C
+napi_status status = napi_status_generic_failure;
+
+// var obj = {}
+napi_value obj, value;
+status = napi_create_object(env, &obj);
+if (status != napi_ok) return status;
+
+// Create a napi_value for 123
+status = napi_create_number(env, 123, &value);
+if (status != napi_ok) return status;
+
+// obj.myProp = 123
+status = napi_set_named_property(env, obj, "myProp", value);
+if (status != napi_ok) return status;
+```
+
+You can also set index-properties in a similar manner. Consider the following
+JavaScript snippet:
+```js
+var arr = [];
+arr[123] = 'hello';
+```
+You can do the equivalent using N-API values with the following snippet:
+```C
+napi_status status = napi_status_generic_failure;
+
+// var arr = [];
+napi_value arr, value;
+status = napi_create_array(env, &arr);
+if (status != napi_ok) return status;
+
+// Create a napi_value for 'hello'
+status = napi_create_string_utf8(env, "hello", -1, &value);
+if (status != napi_ok) return status;
+
+// arr[123] = 'hello';
+status = napi_set_element(env, arr, 123, value);
+if (status != napi_ok) return status;
+```
+
+You can also get properties using the APIs described in this section.
+Consider the following JavaScript snippet:
+```js
+var arr = [];
```

Comment:
Let's insert a `var value;` after this line, so we can assign `value` later rather than using `return`, because performing the C equivalent of `return` is unnecessarily complicated for this example. After all, it involves mentioning the `napi_callback_info`.